model.py

```diff
@@ -137,3 +192,246 @@ def forward(self, inputs: Dict[str, th.Tensor]) -> th.Tensor:
+                axis 1 represents coordinates in the order of (dx,dy,dz).
+        """
+
+        K: int = self.num_object
+        clf: th.Tensor = inputs['clf']  # ..., K, [... volume ...]
+        txn: th.Tensor = inputs['txn']  # ..., 3
+        rxn: th.Tensor = inputs['rxn']  # ..., 3x3
+        device = clf.device
+
+        # Occupancy probability of each object instance.
+        # NOTE(ycho): K-1, since the last channel is reserved for `background`.
+        clf_object = th.narrow(clf, 1, 0, K - 1)  # clf[:, :K-1]
+        voxel_count = einops.reduce(clf_object, 'b n ... -> b n', 'sum')
+        # Sum of coordinates weighted by occupancy probabilities.
+        numer = th.einsum(
+            'b k ..., d ... -> b k d',
+            clf_object,
+            self.xyz.to(device, dtype=clf_object.dtype))
+        denom = th.unsqueeze(voxel_count, dim=2)
+        # NOTE(ycho):
+        # `piv` is roughly the center of mass.
+        piv = numer / denom
+
+        # Add identity transforms at the end of the channel,
+        # B (K-1) ... -> B K ...
+        txn0 = th.zeros_like(txn[:, :1])
+        txn = th.cat([txn, txn0], dim=1).unsqueeze(-1)
+        rxn0 = einops.repeat(th.eye(3, dtype=rxn.dtype),
+                             'i o -> b k i o', b=rxn.shape[0], k=1)
+        rxn = th.cat([rxn, rxn0], dim=1)
+        piv0 = th.zeros_like(piv[:, :1])
+        piv = th.cat([piv, piv0], dim=1).unsqueeze(-1)
+
+        xyz = einops.repeat(self.xyz, 'd x y z -> b k d (x y z)',
+                            b=rxn.shape[0], k=K)
+        xyz2 = rxn @ (xyz - piv) + piv + txn
+        motion = (xyz2 - xyz).view(*xyz.shape[:3], *self.xyz.shape[-3:])
+        motion = th.sum(motion * th.unsqueeze(clf, 2), 1)
+        return motion
+
+
+class SO3(nn.Module):
+    """Euler angle parameterization of the rotation matrix."""
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x: th.Tensor) -> th.Tensor:
+        """Rotation matrix from euler angle parameterization.
+
+        Args:
+            x: Euler angles of shape (..., 3).
+                (roll,pitch,yaw) convention.
+        Returns:
+            Rotation matrix of shape (..., 3, 3),
+                R@x convention.
+        """
+
+        c, s = th.cos(x), th.sin(x)
+        cx, cy, cz = [c[..., i] for i in range(3)]
+        sx, sy, sz = [s[..., i] for i in range(3)]
+
+        # NOTE(ycho):
+        # Represents Rz@Ry@Rx matrix,
+        # Which is equivalent to:
+        # rotate_yaw(rotate_pitch(rotate_roll(point))).
+        r = th.stack([
+            cy * cz,
+            (sx * sy * cz) - (cx * sz),
+            (cx * sy * cz) + (sx * sz),
+            cy * sz,
+            (sx * sy * sz) + (cx * cz),
+            (cx * sy * sz) - (sx * cz),
+            -sy,
+            (sx * cy),
+            (cx * cy)], dim=-1)
+
+        r = einops.rearrange(r, '... (i o) -> ... i o', i=3, o=3)
+        return r
+
+
+class MotionPredictor(nn.Module):
```

Comment:
This class corresponds to https://github.com/columbia-ai-robotics/dsr/blob/f34d60f885cd01e6b562e799d7c81eafda3ae765/model.py#L309-L330, right?

Reply:
yes.